File: solutions/elixir/square-root/3/lib/square_root.ex
defmodule SquareRoot do
  @doc """
  Calculate the integer square root of a positive integer
  """
  @spec calculate(radicand :: pos_integer) :: pos_integer
  def calculate(radicand) do
    1..radicand
    |> Stream.take_while(&(&1 * &1 <= radicand))
    |> Enum.to_list()
    |> List.last()
  end
end

File: solutions/lua/square-root/3/square-root.lua
local SquareRoot = {}

function SquareRoot.square_root(radicand)
	local step = function(n)
		return (n + radicand / n) / 2
	end

	local x0 = radicand / 2
	local x1 = step(x0)

	while x1 < x0 do
		x0 = x1
		x1 = step(x0)
	end

	return math.floor(x0 + 0.5)
end

return SquareRoot
